src/System.Runtime/tests/System/Reflection/TypeInfoTests.cs

```diff
@@ -0,0 +1,223 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.IO;
+using System.Linq;
+using Xunit;
+
+namespace System.Reflection.Tests
+{
+    public class TestType : Stream
+    {
+        public TestType()
+        {
+        }
+
+        public class Nested
+        {
+
+        }
+
+#pragma warning disable 0067 // event never used
+        public event Action<int> StuffHappened;
+#pragma warning restore 0067
+
+#pragma warning disable 0169 // field never used
+        private int _pizzaSize;
+#pragma warning restore 0169
+
+        public override bool CanRead => false;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length
+        {
+            get
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        public override long Position
+        {
+            get
+            {
+                throw new NotImplementedException();
+            }
+
+            set
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        public override void Flush()
+        {
+            throw new NotImplementedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    public class TypeInfoTests
+    {
+        public TypeInfo TypeInfo => typeof(TestType).GetTypeInfo();
+
+        [Fact]
+        public void DeclaredConstructors()
+        {
+            var ctors = TypeInfo.DeclaredConstructors.ToList();
+            Assert.Equal(1, ctors.Count);
+            Assert.True(ctors[0].IsSpecialName);
+            Assert.Equal(0, ctors[0].GetParameters().Count());
+        }
+
+        [Fact]
+        public void DeclaredEvents()
+        {
+            var events = TypeInfo.DeclaredEvents.ToList();
+            Assert.Equal(1, events.Count);
+            Assert.Equal("StuffHappened", events[0].Name);
+            Assert.Equal("Action`1", events[0].EventHandlerType.Name);
+        }
+
+        [Fact]
+        public void GetDeclaredEvent()
+        {
+            Assert.Equal("StuffHappened", TypeInfo.GetDeclaredEvent("StuffHappened").Name);
+        }
+
+        [Fact]
+        public void DeclaredFields()
+        {
+            var fields = TypeInfo.DeclaredFields.ToList();
+            Assert.Equal(2, fields.Count);
+            Assert.Equal("StuffHappened", fields[0].Name);
```

Comment:
You're taking a dependency on the order in DeclaredFields returns fields - there's no guarantee of order.

This applies to all the Type apis that enumerate members on types. It's probably better to do an order-agnostic set compare.